spec/locale.html

```diff
@@ -377,10 +391,9 @@ <h1>
       <dl class="header">
       </dl>
       <emu-alg>
-        1. Assert: _locale_ can be matched by the <code>unicode_locale_id</code> Unicode locale nonterminal.
-        1. Let _languageId_ be the longest prefix of _locale_ matched by the <code>unicode_language_id</code> Unicode locale nonterminal.
-        1. Assert: The first subtag of _languageId_ can be matched by the <code>unicode_language_subtag</code> Unicode locale nonterminal.
-        1. Return the first subtag of _languageId_.
+        1. Let _baseName_ be GetLocaleBaseName(_locale_).
```

Comment:
Is there a reason why the calls to GetLocaleBaseName can't occur in the associated getters rather than in GetLocaleLanguage, GetLocaleScript, etc., in the interest of not calling it twice when calling these AOs from UpdateLanguageId?

Reply:
The call to `GetLocaleBaseName` is spec-only and won't actually happen in implementations. Therefore it seems nicer to allow any `unicode_locale_id` input for `GetLocale{Language,Script,Region}` instead of restricting the input to `unicode_language_id`. (Also I'd prefer if we have fewer references to `unicode_language_id`, because it's too easy to confuse with `unicode_language_subtag`.) 

I'd even prefer if `UpdateLanguageId` calls `GetLocale{Language,Script,Region}` with `tag` instead of `baseName`/`languageId`, because first extracting the base name parts seems like a spec-only optimisation which doesn't affect implementations anyway. For example compare:
- https://github.com/WebKit/WebKit/blob/95b06f4446ccccd849b94e81168963f7cb7ceb2e/Source/JavaScriptCore/runtime/IntlLocale.cpp#L266-L288
- https://github.com/v8/v8/blob/94e627d588d941d5e454adb4296219056a7c4fc8/src/objects/js-locale.cc#L288-L340
- https://github.com/mozilla/gecko-dev/blob/e942f7bc56cd103bba86b396ddeba5b1ab04f1a4/js/src/builtin/intl/Locale.cpp#L314-L390

No implementation actually calls `GetLocale{Language,Script,Region}`, because that requires additional string allocations when compared to only updating the relevant subtags when an option is not `undefined`.